### PR TITLE
REL: automate wheels publication to PyPI

### PR DESCRIPTION
**Pull request by neutrinoceros**
## PR Summary
We've been polishing this workflow on yt-astro-analysis and it's now working correctly so here's the port back to yt itself.
This workflow requires that a secret named "PYPI_TOKEN" be added to the repo (the name is arbitrary, that's just the name we've been using for yt-astro), this can be done by a PyPI maintainer.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
CI is unstable for unrelated reasons, that are adressed in #3760

**Comment by neutrinoceros** [comment]
(rebased on top of #3760)

**Comment by matthewturk** [comment]
This looks good to me, but I can't be the only one to look at it.

**Comment by neutrinoceros** [comment]
@brittonsmith 's review is an absolute requirement as far as I'm concerned

**Comment by neutrinoceros** [comment]
We decided to keep uploading wheels manually for 4.0.2, but I'll add this to the next milestone :)

File: .github/workflows/wheels.yaml [hunk 1]
```diff
@@ -40,4 +42,33 @@ jobs:
...
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+
+      - name: Build sdist
+        run: pipx run build --sdist
+
+      - uses: actions/upload-artifact@v2
+        with:
+          path: dist/*.tar.gz
```
Comment (brittonsmith) [inline .github/workflows/wheels.yaml:58]: Not sure if this is necessary, but in yt_astro_analysis we gave this an explicit name that we used for the download action.
```suggestion
          name: tarball
          path: dist/*.tar.gz
```

File: .github/workflows/wheels.yaml [hunk 2]
```diff
@@ -40,4 +42,33 @@ jobs:
...
+
+  upload_pypi:
+    needs: [build_wheels, build_sdist]
+    runs-on: ubuntu-latest
+    # upload to PyPI on every tag starting with 'yt-'
+    if: github.event_name == 'push' && startsWith(github.event.ref, 'refs/tags/yt-')
+    steps:
+      - uses: actions/download-artifact@v2
+        with:
+          name: artifact
```
Comment (brittonsmith) [inline .github/workflows/wheels.yaml:68]: To get the source tarball
```suggestion
          name: tarball
```

File: .github/workflows/wheels.yaml [hunk 3]
```diff
@@ -40,4 +42,33 @@ jobs:
...
+          path: dist
+
```
Comment (brittonsmith) [inline .github/workflows/wheels.yaml:70]: We also needed a second download for the wheels.
```suggestion
    steps:
      - uses: actions/download-artifact@v2
        with:
          name: wheels
          path: dist
```

**Review by brittonsmith — COMMENTED** [review]
I mostly just compared this to what we now have for yt_astro_analysis.

**Comment by neutrinoceros** [comment]
Thanks a lot @brittonsmith, indeed I made a couple mistakes. I went back and reworked my way there and found the exact them changes you suggested were needed. It should be good now.

**Comment by neutrinoceros** [comment]
taking that back, seems like we just broke the yaml file with invalid syntax, I'll push again (and maybe add a pre-commit hook to validate yaml files so it's easier to detect such mistakes in the future)

**Review by brittonsmith — APPROVED** [review]
This looks ok to me now.

**Comment by neutrinoceros** [comment]
@matthewturk how about we move this to the 4.0.3 milestone ?

**Comment by matthewturk** [comment]
Sounds good.

**Comment by matthewturk** [comment]
I think we still need the `PYPI_SECRET`, right?

**Comment by neutrinoceros** [comment]
correct

**Comment by matthewturk** [comment]
I've added `PYPI_TOKEN` to this repository's secrets.

**Comment by neutrinoceros** [comment]
I reaction to Github bumping `"windows-latest"` 3yrs forward sometime this week, I pushed another commit to pin exact versions of the images we use for Windows and MacOS. Building on old OSes maximise forward compatibility (this is the trick used to make Linux wheels as universal as possible with the ManyLinux docker image). We also do this already in yt_astro_analysis

**Review by brittonsmith — APPROVED** [review]

**Comment by neutrinoceros** [comment]
back ported manually into #3789

**Pull request merged**